flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/v2/FlinkSink.java

```diff
@@ -0,0 +1,603 @@
+      Preconditions.checkState(
+          !flinkWriteConf.overwriteMode(),
+          "OVERWRITE mode shouldn't be enable when configuring to use UPSERT data stream.");
+      Preconditions.checkState(
+          !equalityFieldIds.isEmpty(),
+          "Equality field columns shouldn't be empty when configuring to use UPSERT data stream.");
+      if (!table.spec().isUnpartitioned()) {
+        for (PartitionField partitionField : table.spec().fields()) {
+          Preconditions.checkState(equalityFieldIds.contains(partitionField.sourceId()),
+              "In UPSERT mode, partition field '%s' should be included in equality fields: '%s'",
+              partitionField, equalityFieldColumns);
+        }
+      }
+    }
+
+    this.upsertMode = flinkWriteConf.upsertMode();
+    this.overwrite = flinkWriteConf.overwriteMode();
+    this.distributionMode = flinkWriteConf.distributionMode();
+    this.workerPoolSize = flinkWriteConf.workerPoolSize();
+    this.targetDataFileSize = flinkWriteConf.targetDataFileSize();
+    this.dataFileFormat = flinkWriteConf.dataFileFormat();
+  }
+
+  @Override
+  public DataStream<RowData> addPreWriteTopology(DataStream<RowData> inputDataStream) {
+    return distributeDataStream(inputDataStream, equalityFieldIds, table.spec(),
+        table.schema(), flinkRowType, distributionMode, equalityFieldColumns);
+  }
+
+  @Override
+  public IcebergStreamWriter createWriter(InitContext context) {
+    return restoreWriter(context, null);
+  }
+
+  @Override
+  public IcebergStreamWriter restoreWriter(InitContext context,
+                                           Collection<IcebergStreamWriterState> recoveredState) {
+    StreamingRuntimeContext runtimeContext = runtimeContextHidden(context);
+    int attemptNumber = runtimeContext.getAttemptNumber();
+    String jobId = runtimeContext.getJobId().toString();
+    IcebergStreamWriter streamWriter = createStreamWriter(table, targetDataFileSize, dataFileFormat, upsertMode,
+        flinkRowType, equalityFieldIds, jobId, context.getSubtaskId(), attemptNumber);
+    if (recoveredState == null) {
+      return streamWriter;
+    }
+
+    return streamWriter.restoreWriter(recoveredState);
+  }
+
+  @Override
+  public SimpleVersionedSerializer<IcebergStreamWriterState> getWriterStateSerializer() {
+    return new IcebergStreamWriterStateSerializer<>();
+  }
+
+  static IcebergStreamWriter createStreamWriter(Table table,
+                                                long targetDataFileSize,
+                                                FileFormat dataFileFormat,
+                                                boolean upsertMode,
+                                                RowType flinkRowType,
+                                                List<Integer> equalityFieldIds,
+                                                String jobId,
+                                                int subTaskId,
+                                                long attemptId) {
```

Comment:
Having this many parameters are confusing. Can we change this to `private`? Or is this expected to be overwritten?

If we need to keep this, would a javadoc provide some more context?

Reply:
I removed this method.

Reply:
Thanks!